java/core/src/main/java/co/worklytics/psoxy/impl/SanitizerImpl.java

```diff
@@ -91,7 +112,93 @@ public String sanitize(@NonNull URL url, @NonNull String jsonResponse) {
             return jsonResponse;
         }
 
-        //q: move this stuff to initialization / DI provider??
+        if (getOptions().getRules2() == null) {
+            return legacyTransform(url, jsonResponse);
+        } else {
+            return transform(url, jsonResponse);
+        }
+    }
+
+
+    String transform(@NonNull URL url, @NonNull String jsonResponse) {
+        if (compiledEndpointRules == null) {
+            compiledEndpointRules = options.getRules2().getEndpoints().stream()
+                .map(endpoint -> Pair.of(Pattern.compile(endpoint.getPathRegex(), CASE_INSENSITIVE), endpoint))
+                .collect(Collectors.toList());
+        }
+
+        String relativeUrl = URLUtils.relativeURL(url);
+        Optional<Pair<Pattern, Rules2.Endpoint>> matchingEndpoint = compiledEndpointRules.stream()
+            .filter(compiledEndpoint -> compiledEndpoint.getKey().asMatchPredicate().test(relativeUrl))
+            .findFirst();
+
+        return matchingEndpoint.map(match -> {
+
+            Object document = jsonConfiguration.jsonProvider().parse(jsonResponse);
+
+            for (Transform transform : match.getValue().getTransforms()) {
+                document = applyTransform(transform, document);
+            }
+            return jsonConfiguration.jsonProvider().toJson(document);
+        }).orElse(jsonResponse);
+    }
+
+
+    Object applyTransform(Transform transform, Object document ) {
+        List<JsonPath> paths = compiledTransforms.computeIfAbsent(transform,
+            t -> t.getJsonPaths().stream()
+                .map(JsonPath::compile)
+                .collect(Collectors.toList()));
+
+        if (transform instanceof Transform.Redact) {
+            for (JsonPath path : paths) {
+                document = path.delete(document, jsonConfiguration);
+            }
+        } else if (transform instanceof Transform.Pseudonymize) {
+
+            //curry the defaultScopeId from the transform into the pseudonymization method
+            MapFunction f =
+                ((Transform.Pseudonymize) transform).getIncludeOriginal() ? this::pseudonymizeWithOriginalToJson : this::pseudonymizeToJson;
+            for (JsonPath path : paths) {
+                document = path.map(document, f, jsonConfiguration);
+            }
+
+        } else if (transform instanceof Transform.PseudonymizeEmailHeader) {
+            for (JsonPath path : paths) {
+                document = path.map(document, this::pseudonymizeEmailHeaderToJson, jsonConfiguration);
+            }
+        } else if (transform instanceof Transform.RedactRegexMatches) {
+            MapFunction f = getRedactRegexMatches((Transform.RedactRegexMatches) transform);
+            for (JsonPath path : paths) {
+                document = path.map(document, f, jsonConfiguration);
+            }
```

Comment:
i considered that, but without (huge) refactor it would couple the transforms to our implementation (in particular, the `JsonPath` library. 

eg, redact is
```java
            for (JsonPath path : paths) {
                document = path.delete(document, jsonConfiguration);
            }
```
so would need to compile the paths, as well as be aware of the `jsonConfiguration` thing being passed in as argument ...